### PR TITLE
bpo-33332: Add signal.valid_signals()

### PR DESCRIPTION
**Pull request by pitrou**
https://bugs.python.org/issue33332

### CODE REVIEW

File: Lib/test/test_signal.py [hunk 1]
```diff
@@ -67,6 +67,13 @@ def test_interprocess_signal(self):
         script = os.path.join(dirname, 'signalinterproctester.py')
         assert_python_ok(script)
 
+    def test_valid_signals(self):
+        s = signal.valid_signals()
+        self.assertIsInstance(s, set)
+        self.assertIn(signal.Signals.SIGINT, s)
+        self.assertNotIn(0, s)
+        self.assertNotIn(signal.NSIG, s)
```
Comment (vstinner) [inline Lib/test/test_signal.py]: What do you think of testing len(s) < NSIG? Maybe len(s) <= NSIG?

**Review by vstinner — COMMENTED** [review]
It would be interesting to also implement the feature on Windows where "valid signals" is hard to guess for a Unix user. See signal_signal() of Modules/signalmodule.c for valid signals: I count 6 signals, 7 if SIGBREAK is also available.

**Comment by vstinner** [comment]
You may also modify:
* Lib/asyncio/unix_events.py: "if not (1 <= sig < signal.NSIG):" in _check_signal() -- don't forget to patch Lib/test/test_asyncio/test_unix_events.py
* Lib/test/support/__init__.py: SaveSignals uses "self.signals = list(range(1, signal.NSIG))"

File: Lib/test/test_signal.py (hunk 1 above)
Comment (pitrou) [reply Lib/test/test_signal.py]: I'm not sure it's very useful.

**Comment by pitrou** [comment]
Thanks for the comments. You're right about Windows. I'll also fix asyncio and test.support.

File: Lib/test/test_signal.py (hunk 1 above)
Comment (vstinner) [reply Lib/test/test_signal.py]: The purpose is to test if NSIG is consistent with sigfillset(), to check if something is really wrong on a platform. I would like to know if range(0, NSIG) miss some signals on some platforms :-)

**Comment by 1st1** [comment]
Big +1 for this feature. Can you make it return a `frozenset` though?

**Comment by pitrou** [comment]
Frozen sets are nice if you want to hash them, otherwise I don't really see the point.

**Comment by 1st1** [comment]
> Frozen sets are nice if you want to hash them, otherwise I don't really see the point.

What's the point of being able to modify the returned set? ;) If you return a frozenset you can cache it (something that I would recommend doing in this PR anyways).

This isn't a deal breaker, it's just that I would use a frozen set here. Overall LGTM.

**Comment by pitrou** [comment]
Other functions like `signal.pthread_sigmask` and `signal.sigpending` return plain sets. So we could return a frozenset from `signal.valid_signals` but that sounds a bit gratuitous. @vstinner, what do you think?

**Comment by vstinner** [comment]
First I wanted to get a list, but then I read the PR and I like reusing existing code :-)

If someone wants an immutable type, it's very simple: frozenset(signal.valid_signals()).

An immutable type might be justified if it was a module constant, but it's a function which creates a new object at each call, so IMHO a set is just fine.

I don't want to modify other signal functions just to be pedantic. I like Antoine's rationale to use set in valid_signals().

**Pull request merged**